File: python/src/payskill/signer.py
"""Signer interface and implementations for the pay SDK.

Three modes:
1. CLI signer (default): subprocess call to `pay sign`
2. Raw key: from PAYSKILL_KEY environment variable (dev/testing only)
3. Custom: user provides a sign(hash) -> signature callback
"""

from __future__ import annotations

import os
import subprocess
from abc import ABC, abstractmethod
from collections.abc import Callable

from eth_account import Account


class Signer(ABC):
    """Abstract signer interface."""

    @property
    @abstractmethod
    def address(self) -> str:
        """The signer's Ethereum address (0x-prefixed, checksummed)."""

    @abstractmethod
    def sign(self, hash_bytes: bytes) -> bytes:
        """Sign a 32-byte hash and return 65-byte signature (r || s || v)."""


class CliSigner(Signer):
    """Signs via the `pay sign` CLI subprocess."""

    def __init__(self, command: str = "pay", address: str = "") -> None:
        self.command = command
        self._address = address
        if not self._address:
            try:
                result = subprocess.run(
                    [self.command, "address"],
                    capture_output=True,
                    text=True,
                    timeout=10,
                )
                if result.returncode == 0:
                    self._address = result.stdout.strip()
            except Exception:
                self._address = ""

    @property
    def address(self) -> str:
        return self._address

    def sign(self, hash_bytes: bytes) -> bytes:
        result = subprocess.run(
            [self.command, "sign"],
            input=hash_bytes.hex(),
            capture_output=True,
            text=True,
            timeout=30,
        )
        if result.returncode != 0:
            msg = f"CLI signer failed: {result.stderr.strip()}"
            raise RuntimeError(msg)
        return bytes.fromhex(result.stdout.strip())


class RawKeySigner(Signer):
    """Signs with a raw private key using eth_account. Dev/testing only."""

    def __init__(self, key: str | None = None) -> None:
        self._key = key or os.environ.get("PAYSKILL_KEY", "")
        if not self._key:
            msg = "No key provided and PAYSKILL_KEY not set"
            raise ValueError(msg)
        self._account = Account.from_key(self._key)

    @property
    def address(self) -> str:
        return str(self._account.address)

    def sign(self, hash_bytes: bytes) -> bytes:
        """Sign a 32-byte hash with ECDSA. Returns 65-byte r||s||v signature."""
        signed = self._account.unsafe_sign_hash(hash_bytes)
        # Construct 65-byte signature: r (32 bytes) || s (32 bytes) || v (1 byte)
        r_bytes = int(signed.r).to_bytes(32, "big")
        s_bytes = int(signed.s).to_bytes(32, "big")
        v_bytes = bytes([int(signed.v)])
        return bytes(r_bytes + s_bytes + v_bytes)


class CallbackSigner(Signer):
    """Delegates signing to a user-provided callback."""

    def __init__(
        self,
        callback: Callable[[bytes], bytes],
        address: str = "",
    ) -> None:
        self._callback = callback
        self._address = address

    @property
    def address(self) -> str:
        return self._address

    def sign(self, hash_bytes: bytes) -> bytes:
        return self._callback(hash_bytes)


def create_signer(mode: str = "cli", **kwargs: object) -> Signer:
    """Factory for creating signers.

    Args:
        mode: "cli" (default), "raw", or "custom"
        **kwargs: Passed to the signer constructor.
            - cli: command (str), address (str)
            - raw: key (str)
            - custom: callback (Callable[[bytes], bytes]), address (str)
    """
    if mode == "cli":
        return CliSigner(
            command=str(kwargs.get("command", "pay")),
            address=str(kwargs.get("address", "")),
        )
    if mode == "raw":
        return RawKeySigner(key=kwargs.get("key"))  # type: ignore[arg-type]
    if mode == "custom":
        callback = kwargs.get("callback")
        if callback is None:
            msg = "custom signer requires a callback"
            raise ValueError(msg)
        return CallbackSigner(
            callback=callback,  # type: ignore[arg-type]
            address=str(kwargs.get("address", "")),
        )
    msg = f"Unknown signer mode: {mode}"
    raise ValueError(msg)

File: typescript/src/signer.ts
/**
 * Signer interface and implementations.
 *
 * Three modes:
 * 1. CLI signer (default): subprocess call to `pay sign`
 * 2. Raw key: from PAYSKILL_KEY environment variable (dev/testing only)
 * 3. Custom: user provides a sign(hash) -> signature callback
 */

import { execFileSync } from "node:child_process";
import { type Hex } from "viem";
import {
  privateKeyToAccount,
  sign as viemSign,
  serializeSignature,
} from "viem/accounts";

/** Abstract signer interface. */
export interface Signer {
  /** Sign a 32-byte hash. Returns 65-byte signature (r || s || v). */
  sign(hash: Uint8Array): Uint8Array;
  /** The signer's Ethereum address (0x-prefixed, checksummed). */
  readonly address: string;
}

/** Signs via the `pay sign` CLI subprocess. */
export class CliSigner implements Signer {
  private readonly command: string;
  readonly address: string;

  constructor(command = "pay", address = "") {
    this.command = command;
    if (address) {
      this.address = address;
    } else {
      try {
        this.address = execFileSync(this.command, ["address"], {
          encoding: "utf-8",
          timeout: 10_000,
        }).trim();
      } catch {
        this.address = "";
      }
    }
  }

  sign(hash: Uint8Array): Uint8Array {
    const hexHash = Buffer.from(hash).toString("hex");
    const result = execFileSync(this.command, ["sign"], {
      input: hexHash,
      encoding: "utf-8",
      timeout: 30_000,
    });
    return Buffer.from(result.trim(), "hex");
  }
}

/** Signs with a raw private key using viem. Dev/testing only. */
export class RawKeySigner implements Signer {
  private readonly _key: Hex;
  readonly address: string;

  constructor(key?: string) {
    const rawKey = key ?? process.env.PAYSKILL_KEY ?? "";
    if (!rawKey) {
      throw new Error("No key provided and PAYSKILL_KEY not set");
    }
    this._key = rawKey.startsWith("0x")
      ? (rawKey as Hex)
      : (`0x${rawKey}` as Hex);
    const account = privateKeyToAccount(this._key);
    this.address = account.address;
  }

  sign(_hash: Uint8Array): Uint8Array {
    // viem's sign() does raw ECDSA signing (no EIP-191 prefix)
    // sign() is async but Signer interface is sync — we use the sync internal
    // For the sync interface, we need a workaround. Since RawKeySigner is
    // dev/testing only, we'll do a sync computation via signSync.
    // viem doesn't expose sync sign, but we can construct manually.
    // Use the async path via signAsync helper pattern.
    throw new Error(
      "RawKeySigner.sign() is synchronous but viem signing is async. " +
        "Use RawKeySigner.signAsync() or use buildAuthHeaders() directly with the private key."
    );
  }

  /** Async sign — preferred over sync sign(). */
  async signAsync(hash: Uint8Array): Promise<Uint8Array> {
    const hashHex = ("0x" + Buffer.from(hash).toString("hex")) as Hex;
    const raw = await viemSign({ hash: hashHex, privateKey: this._key });
    const sigHex = serializeSignature(raw);
    return hexToBytes(sigHex);
  }
}

/** Delegates signing to a user-provided callback. */
export class CallbackSigner implements Signer {
  private readonly callback: (hash: Uint8Array) => Uint8Array;
  readonly address: string;

  constructor(
    callback: (hash: Uint8Array) => Uint8Array,
    address = ""
  ) {
    this.callback = callback;
    this.address = address;
  }

  sign(hash: Uint8Array): Uint8Array {
    return this.callback(hash);
  }
}

/** Factory for creating signers. */
export function createSigner(
  mode: "cli" | "raw" | "custom" = "cli",
  options: {
    command?: string;
    key?: string;
    address?: string;
    callback?: (hash: Uint8Array) => Uint8Array;
  } = {}
): Signer {
  switch (mode) {
    case "cli":
      return new CliSigner(options.command, options.address);
    case "raw":
      return new RawKeySigner(options.key);
    case "custom":
      if (!options.callback) {
        throw new Error("custom signer requires a callback");
      }
      return new CallbackSigner(options.callback, options.address);
    default:
      throw new Error(`Unknown signer mode: ${mode as string}`);
  }
}

function hexToBytes(hex: string): Uint8Array {
  const clean = hex.startsWith("0x") ? hex.slice(2) : hex;
  const bytes = new Uint8Array(clean.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(clean.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}
